### ISSUE TITLE
Progress bar 

### ISSUE THREAD

**Issue by jgabry** [stan-dev/cmdstanr#18]
CmdStanPy has a nice progress bar that we can implement in R too. 

We are currently using `processx::run(..., echo=TRUE)` so that the R user sees the output directly from CmdStan, but processx provides a lot more control than that and it looks like it has what we’d need behind the scenes for a progress bar even for chains in parallel.

**Comment by jgabry** [comment]
After discussing with @mitzimorris I found out that maybe there are some problems with the progress bar implementation inn CmdStanPy so maybe we shouldn't try to copy that until it's sorted out there. 

Also, an alternative to a progress bar would be to run background processes and provide a way for the user to check the status. A progress bar isn't necessary for models that run fast, and maybe for slow models it's better to let them just run in the background and allow the user to check the status when they want? This is possible with the processx package we're already using. 

Anyway, I haven't thought too hard about this yet so I'm definitely open to other ideas!

**Comment by rok-cesnovar** [comment]
So a progress bar would be fairly simple to implement after #61 is merged, provided that a user doesnt specify refresh=0. @jgabry do we still want to do this?

**Comment by jgabry** [comment]
I could go either way on this. The current output from parallelization is fine, although a progress bar could be nicer. What do you think?

**Comment by rok-cesnovar** [comment]
Do you have an example for an R package that uses progress bars? I am not completely sure how it would look. Worth a try tho.

**Comment by jgabry** [comment]
Yeah it's worth a try. Things to check out whenever you want to dive into this (definitely not super urgent though): 

* I haven't tried this, but maybe it has what we need: `progress` package (https://github.com/r-lib/progress, https://cran.r-project.org/package=progress). Has some of same very capable authors as the `processx` package that we're using. 
* I've used the `revdepcheck` package (https://github.com/r-lib/revdepcheck, for testing reverse dependencies) and that has a really nice progress bar. Looks like it uses the `progress` package.
* You can also look at what other packages use `progress` by looking at it's CRAN page: https://cran.r-project.org/package=progress (scroll down to reverse dependencies section). There are a bunch so you may be able to look at those packages for implementation details. 

There are other packages for making progress bars but I haven't used them. I've heard it can be difficult to do progress bars for parallelized code in R, but I haven't gone down that rabbit hole so I don't know any details about what the challenges are unfortunately.

**Comment by mike-lawrence** [comment]
What do folks think of the approach taken by my package [ezStan](https://github.com/mike-lawrence/ezStan) where I simply watch the contents of the CSV files to establish progress (inc. reporting post-warmup divergences and permitting termination on detection of these)?

**Comment by rok-cesnovar** [comment]
@mike-lawrence that sounds like a great idea. 

So you essentially count the number of lines in the CSV? Do you read the entire CSV in a separate thread constantly for checking divergences?

**Comment by mike-lawrence** [comment]
> So you essentially count the number of lines in the CSV? Do you read the entire CSV in a separate thread constantly for checking divergences?

Precisely. My package is a bit of a monstrosity of poorly-commented/hack-y code, so if this kind of progress indicator is of interest for cmdstanr I'd probably start from scratch. But it would be neat to finally contribute to the project beyond my forum activity, so I'm keen to do it.

**Comment by rok-cesnovar** [comment]
Cool! I am definitely in favor.

Another option that would only allow showing progress (not divergences and other diagnostic realtime stuff) is to rely on stdout. If the progress bar is enabled then set refresh = 1. That means that we get information for each iteration. 

If a user wants realtime reports for divergence/treedepth then we need to read the CSV.

**Comment by mike-lawrence** [comment]
Good idea, not monitoring the csv if the user hasn't asked for any csv-content-dependent behaviour will make things more resource-friendly during sampling. I'll have to teach myself how to monitor the stdout instead, but I suspect that's within my capacity.

**Comment by rok-cesnovar** [comment]
For sampling we have a process function with a "state machine" with which we don't do much, mostly just limit that the initial Cmdstan metadata dump is not printed and to mark the transition from warmup to sampling and such: https://github.com/stan-dev/cmdstanr/blob/master/R/run.R#L650
This is called every 200ms I believe if any stdout came in.

For all the other methods the function is here https://github.com/stan-dev/cmdstanr/blob/master/R/run.R#L791
Here we only make sure the Cmdstan metadata dump is not printed.

**Comment by jgabry** [comment]
> But it would be neat to finally contribute to the project beyond my forum activity, so I'm keen to do it.

Sounds great!

**Comment by mike-lawrence** [comment]
Just a heads up that I'll start working on this today. I'd somehow gotten it into my head that I was waiting on some info but see that all I need is in Rok's last comment. I'll start with the simple progress bar that watches stdout, then once that's working solidly dive into the csv-informed monitor. 

Tangentially, now that I've seen that the ArViz crew is using HDF5/NetCDF, I'm more confident that that's at least not a completely ridiculous thing to work on as a replacement for csv in cmdstan, so I may look at that in parallel to the monitor (which would benefit something that's not csv).

**Comment by rok-cesnovar** [comment]
Great, let me know if you need any additional info or help. Thanks!

**Comment by mike-lawrence** [comment]
Just an update on this that finally dove into it this weekend and after getting up to speed on R6 I have something partially working [here](https://github.com/mike-lawrence/cmdstanr). I'll do a pull request when it's actually ready. I ended up not being able to use any of the existing progress bar packages because none permit multiple simultaneous progress bars, and after getting my own multi-line updates seemingly-working I [discovered why](https://stackoverflow.com/questions/63985984/). If I don't find a workaround for that snag, I'll implement a minimal all-chains-on-one-line output that will be shown when multi-line isn't going to work.

**Comment by mike-lawrence** [comment]
Getting [closer](https://github.com/mike-lawrence/cmdstanr). Here's what I was hoping to get working in terms of multi-line indicators: 
![ezgif-6-0a265010915b](https://user-images.githubusercontent.com/150781/93929115-bc9dd680-fcf1-11ea-9075-f508c83ad58b.gif)
(n.b. the top bar indicates warmup vs sampling periods and the `?` during warmup indicates large uncertainty in the time remaining thanks to warmup times not really being very predictive of sampling times; the eta computation during sampling uses only the times during sampling, so have much less uncertainty and therefore no `?`)

But it turns out that: 
1. Terminals (at least the ones I've tried) don't seem to respect the '\r' character when text has wrapped onto new lines, yielding '\n' behaviour instead
2. Terminals (at least the ones I've tried) don't seem to consistently set `options('width')`; workaround on unix-alikes is to use `tput cols`, and while this value updates if you resize the window, it doesn't update if you resize the window while `mod$sample()` is running. This latter isn't a big deal, but might cause confusion for users if they expect things to resize nicely at all times.
3. RStudio's console *does* update `options('width')` even while `mod$sample()` is running, but seemingly with very fuzzy logic such that resizing the window just a little bit can change the output wrapping behaivour but not the value of `options('width')`. Again not a huge deal.
4. RStudio's console's font rendering is pretty shoddy, making for less pretty output there than in the gif above.

While I've put a lot of work into the multi-line solution (even using the unicode partial-block characters for finer-than-single-character progress indication), given 1 I should probably just implement a single-line progress along the lines of:
`[1: 85% 5m]  [2: 87% 4m]  [3: 83% 6m] [4: 85% 4m] `

Maybe I can have a check for if there is enough room to add progress bars for each process. 

Another idea is to not use the console/terminal at all and instead pop out a tcltk progress window or something?

**Comment by mike-lawrence** [comment]
Actually, now that I think of it, I should check how cmdstanpy does their progress bars to see if they encountered and worked around the blocks I'm experiencing. @mitzimorris whereabouts would I look for the progress bar code in the cmdstanpy source?￼

**Comment by mike-lawrence** [comment]
Oh! I just discovered that the terminal emulators that don't respect the `\r` character *do* respect the `\033[F` character! I'll still add a one-line option, but psyched that I don't have to abandon the nicer multi-line format.

**Comment by rok-cesnovar** [comment]
Nice!

**Comment by mike-lawrence** [comment]
Belatedly getting back to this. The in-console progress is pretty much done, and I'll do a PR for that shortly, but I thought I'd get folks thoughts here (I didn't feel it necessary to create a whole new issue for this) on an idea for a subsequent PR: 

Given the likely substantial proportion of the user base that is using Rstudio, I wonder if it makes sense to work on an option to use the Rstudio "jobs" feature where each chain would go in it's own job. Frankly, if I'd thought of this earlier, I wouldn't have bothered with the console progress bars as the jobs system has all sorts of nifty progress stuff built in (bars, status indicators, etc.), making for lighter code on cmdstanr's side. Using the jobs system also has the side-benefit of not blocking the rsession, permitting users to do other stuff while the chains are churning.

**Comment by jgabry** [comment]
I like the idea of using the jobs system, but I'm not super familiar with it. From what you describe it sounds great. Any downsides you can think of?

**Comment by mike-lawrence** [comment]
Really simply that it's a feature that is only useful to those users using Rstudio. I'll go ahead and finish the in-console progress indicators, then tackle the jobs idea separately.

**Comment by mike-lawrence** [comment]
Oh! I just realized a pretty important downside to using rstudio jobs; there would either be only one progress bar or potentially lots of overhead associated with multiple rsesions each watching for output from independent runs. Do we have a sense as to how much compute time `mod$sample()` consumes from watching the output of `processx::run()`? Is it a "output triggers computation" thing, or does it have to constantly check in for new output?

**Comment by jgabry** [comment]
Yeah we do need to check regularly for new output. Inside a while loop we have

https://github.com/stan-dev/cmdstanr/blob/962e7c1dcd2b779f75d36ea2ed35f5e2fcc639b5/R/run.R#L274-L275

where `procs$poll` calls `processx::poll` (https://processx.r-lib.org/reference/poll.html)

**Comment by mike-lawrence** [comment]
So I'm reconsidering this consequent to @mitzimorris' reasonable reminders that it is valuable to keep cmdstanr light. That is, since the approach to progress bars I've been implementing simply looks at output that is otherwise already printed, it would be possible for me to move said bars into an entirely different package. A con to doing the bars in a separate package is that then there would be some extra overhead from two levels of checking the cmdstan outputs rather than one. Thoughts?

**Comment by jgabry** [comment]
Yeah good point, I guess that could certainly go in a separate package that cmdstanr can suggest. That would be fine by me. 

> A con to doing the bars in a separate package is that then there would be some extra overhead from two levels of checking the cmdstan outputs rather than one.

If it's possible to evaluate and document what the cost roughly is (which should be doable) then I think it's OK. For most users it would probably still be worth using the progress bar even with a bit of extra overhead, but if the extra little bit of speed really matters for someone then they can just not use the progress bar.

**Comment by mike-lawrence** [comment]
After thinking on this, it'll also be frankly easier on my end in a few respects too. What I think I'll do is revamp my [ezStan](https://github.com/mike-lawrence/ezStan) package to use cmdstanr rather than rstan. I'll post back here when the conversion is done and folks can start using it.

**Comment by jgabry** [comment]
Cool, thanks Mike!

**Comment by jgabry** [comment]
Thanks for looking into this. So are you envisioning an argument to the `sample` method that defaults to FALSE but can be turned on if the user has `progressr` installed? That sounds OK to me. How complicated would this be to add? Will anything important need to be rewritten to accommodate this?

**Comment by jgabry** [comment]
One question I forgot to ask before: would the idea be to replace the printed iteration updates with the progress bar or to display both? 

> I'm hoping I might be able to add this fairly easily if I don't run into anything unexpected.
> 

Famous last words ;) 

> Just to note that the way that `progressr` works means that by default simply enabling the progress functionality won't do anything. It will be up to the user to register a handler and define the progress bar.
> 

Ok yeah that's fine, we can just mention that in the documentation for the new argument. 

> The main issue right now is that, sadly, R's progress bar ecosystem mainly doesn't allow multiple progress bars at a time, because RStudio's console doesn't support the needed ANSI codes, so I can't make per-chain progress bars.
> 

That's a shame, but I guess we can add that later if it ever becomes possible. We should also mention in the documentation that the progress is overall, not per chain. 

> I'll have a go and see if I can do it neatly. (At the moment the main issue is parsing the arguments so that I can work out how many iterations/chains/stdout lines per iteration are being requested. I don't think there's anything genuinely hard, though.)

Ok great, thanks! Let me know if you run into any trouble.

**Comment by josswright** [comment]
> One question I forgot to ask before: would the idea be to replace the printed iteration updates with the progress bar or to display both?

My plan was to leave that up to the end user. The `show_stdout_messages` and `show_stderr_messages` arguments are already there, and various handlers can interact with stdout messages in different ways. (I don't use RStudio personally but, for example, there is a handler that puts progress updates directly into RStudio's jobs pane rather than the console.)

It is part of the design of `progressr`, though, that progress reporting shouldn't interfere with normal messages from `print()`, `cat()`, `message()`, etc: https://progressr.futureverse.org/#use-regular-output-as-usual-alongside-progress-updates

At the moment, I'm just planning to hook in the basic progress functionality, and would personally use the existing arguments to turn off printed iteration updates. If there's a chance to make it more flexible in the future--especially if support for nested/multiline progress bars emerges--I'll happily keep at it.

> > I'm hoping I might be able to add this fairly easily if I don't run into anything unexpected.
> 
> Famous last words ;)

I've jinxed it now. :)

> Ok great, thanks! Let me know if you run into any trouble.

Will do! I'll post back here if and when I get it working.

**Comment by jgabry** [comment]
Sounds good, thanks!

**Comment by josswright** [comment]
> Very cool! Thanks for working on this. Some comments:
> 
> > * `show_progress_bar`: Default =`FALSE`.
> >   If `TRUE`, registers a progress bar via `progressr` and signals an update for every output line that matches `"Iteration:"`. The user is responsible for registering those progress updates with an appropriate handler. (In the video below I'm using `cli`'s progress bar, but there are quite a few [alternatives](https://progressr.futureverse.org/articles/progressr-11-handlers.html), including RStudio's jobs pane or even audio updates via `beepr`.)
> 
> This is great. In the doc for this argument we can also point users to the alternatives you mentioned here. I think it's nice to have all these different options even if it means they need to register it themselves.

Thanks. :)

I suppose it might be possible to have a convenience `register_cmdstanr_progress()` function to set up a default handler appropriately formatted for Stan.

> > One thing that might be worthwhile is to report the text of the line causing the update to the progress bar. That would give the end user the option of incorporating the text of the line into the progress bar, for example as a status message, if they feel like it. For the moment, though I wanted to keep things relatively simple. (It's trivial to add this.)
> 
> I'd be curious to see what that looks like. Seems like it could sometimes be useful.

I think it's probably worth passing the information across, even if the user doesn't display it. I'll have a play.

> > Happy to hear comments as to whether this seems like a reasonable approach.
> 
> Do you mind if I post on the Stan forum that you're implementing this and point to this issue where people can give you feedback? I think it would be good to get more people to comment on this design and then also eventually try it out and give feedback once it's ready but before it's merged. (Or you're more than welcome to make the post yourself of course.)

Sure. Happy to see what people think, so please feel free. I'm a little snowed under at the moment, but I'll tidy things up and push it to my fork as soon as I can.
 
> Thanks again for working on this!

Selfish reasons! :)

**Comment by VisruthSK** [comment]
@josswright Hi! I took a look at your fork and it looks good! I have a few small things I'd want to change, but it looks great on the whole. Do you want to open a PR?
